source/lac/trilinos_epetra_vector.cc

```diff
@@ -0,0 +1,529 @@
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LinearAlgebra
+{
+  namespace EpetraWrappers
+  {
+    Vector::Vector()
+      :
+      vector(new Epetra_FEVector(Epetra_Map(0,1,0,Epetra_MpiComm(MPI_COMM_SELF))))
+    {}
+
+
+
+    Vector::Vector(const Vector &V)
+      :
+      vector(new Epetra_FEVector(V.trilinos_vector()))
+    {}
+
+
+
+    Vector::Vector(const IndexSet &parallel_partitioner,
+                   const MPI_Comm &communicator)
+      :
+      vector(new Epetra_FEVector(parallel_partitioner.make_trilinos_map(communicator,false)))
+    {}
+
+
+
+    void Vector::reinit(const IndexSet &parallel_partitioner,
+                        const MPI_Comm &communicator,
+                        const bool      omit_zeroing_entries)
+    {
+      Epetra_Map input_map = parallel_partitioner.make_trilinos_map(communicator,false);
+      if (vector->Map().SameAs(input_map)==false)
+        vector.reset(new Epetra_FEVector(input_map));
+      else if (omit_zeroing_entries==false)
+        {
+          const int ierr = vector->PutScalar(0.);
+          (void) ierr;
+          Assert(ierr==0, ExcTrilinosError(ierr));
+        }
+    }
+
+
+
+    Vector &Vector::operator= (const Vector &V)
+    {
+      // Distinguish three cases:
+      //  - First case: both vectors have the same layout.
+      //  - Second case: both vectors have the same size but different layout.
+      //  - Third case: the vectors have different size.
+      if (vector->Map().SameAs(V.trilinos_vector().Map()))
+        *vector = V.trilinos_vector();
+      else
+        {
+          if (size()==V.size())
+            {
+              Epetra_Import data_exchange(vector->Map(), V.trilinos_vector().Map());
+
+              const int ierr = vector->Import(V.trilinos_vector(), data_exchange, Insert);
+              Assert(ierr==0, ExcTrilinosError(ierr));
+            }
+          else
+            vector.reset(new Epetra_FEVector(V.trilinos_vector()));
+        }
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator= (const ::dealii::LinearAlgebra::ReadWriteVector<double> &V)
+    {
+      IndexSet elements_to_write(V.get_stored_elements());
+      IndexSet locally_owned_elements(this->locally_owned_elements());
+      IndexSet off_proc_elements(elements_to_write);
+      off_proc_elements.subtract_set(locally_owned_elements);
+      IndexSet on_proc_elements(elements_to_write);
+      on_proc_elements.subtract_set(off_proc_elements);
+
+      // Write the elements that locally owned.
+      IndexSet::ElementIterator elem = on_proc_elements.begin(),
+                                elem_end = on_proc_elements.end();
+      for (; elem!=elem_end; ++elem)
+        {
+          const TrilinosWrappers::types::int_type local_index =
+            vector->Map().LID(static_cast<TrilinosWrappers::types::int_type>(*elem));
+          (*vector)[0][local_index] = V[*elem];
+        }
+
+
+      // Write the elements off-processor. Cannot use Import function of Trilinos
```

Comment:
@kronbichler I don't think that creating an ```Epetra_Map``` and a ```Epetra_Import/Export```every time we need to import values will scale. When I said that I wanted to modify ```MPI::Partitioner```, I meant that if we are using Trilinos we would just encapsulate a ```Epetra_Import``` object. Here is the interface, I have in mind:
```c++
MPI::Partitioner part;
EpetraWrappers::Vector epetra_vec;
ReadWriteVector local_vec;
// Fill in local_vec
epetra_vec.import(local_vec, part, VectorOperation::Insert); // Since this is the first time we call import, part creates an epetra_import object.
// Do more calculation on local_vec
epetra_vec.import(local_vec, part, VectorOperation::add); // In debug mode, we have a check that the Epetra_Import object makes sense and then, we reuse it.
```
I looked and it seems that we only need ```Epetra_Import```, we can take care ourselves of the communication for the ```Epetra_Export``` case if we know ```Epetra_Import```. 
I think that separating the communication pattern from the vector itself is the right way to do it. This is similar to what we do with ```SparseMatrix``` and ```SparsityPattern```. This way we can reuse the same communication pattern for vectors that share the same layout. I think this is an important case in practice. Also, we don't need to use a ```MPI::Partitioner``` object, we can create a base class and each library would have its own specialized communication pattern object.